File: src/main/java/sunshine/api/common/ApiResponse.java
package sunshine.api.common;

import java.time.Instant;

public record ApiResponse<T>(boolean success, T data, ErrorResponse error, Instant timestamp) {
    public static <T> ApiResponse<T> ok(T data) {
        return new ApiResponse<>(true, data, null, Instant.now());
    }

    public static <T> ApiResponse<T> error(ErrorResponse error) {
        return new ApiResponse<>(false, null, error, Instant.now());
    }
}

File: src/main/kotlin/com/sunshine/api/response/ApiResponse.kt
package com.sunshine.api.response

data class ApiResponse<T>(
    val success: Boolean,
    val data: T?,
    val error: ErrorResponse?
) {
    companion object {
        fun <T> success(data: T): ApiResponse<T> {
            return ApiResponse(true, data, null)
        }

        fun failure(error: ErrorResponse): ApiResponse<Nothing> {
            return ApiResponse(false, null, error)
        }
    }
}
